### DISCUSSION TITLE
When I use a Union in C, the code generated by bpf2go is not complete.

### DISCUSSION THREAD

**Discussion by binw666** [cilium/ebpf#1041] [Q&A] [answered]
When I use a Union in C, the code generated by bpf2go is not complete. Should I complete it manually? Is there a more elegant way to use it in Go? The C code  is as follows:
```c
typedef struct
{
    __u32 a1;
    __u8 a2;
} A;

typedef struct
{
    __u32 b1;
    __u32 b2;
    __u8 b3;
} B;

typedef struct
{
    __u32 type;
    union value
    {
        A a;
        B b;
    } value;
    
} Value;
```
After compiling, the corresponding part in Go is as follows :
```go
type bpfValue struct {
	Type  uint32
	Value struct {
		A struct {
			A1 uint32
			A2 uint8
			_  [3]byte
		}
		_ [4]byte
	}
}
```
There is only struct A. Completing it manually can easily lead to mistakes. Is there a more elegant way to use it in Go?

**Comment by ti-mo** [comment]
Since Go doesn't have unions, how would you 'complete' this type? The rationale behind the current behaviour is: on the first position in the union ('a' in this case), put the type you want to use to interact with the union from the Go side.

**Comment by binw666** [comment]
Thank you for your response. I know that Go doesn't have unions.What I meant is that not all the C structures have been fully implemented in Go.  It is tedious and error-prone to manually implement these structures in Go. Today, I discovered that I can use `cgo -godefs` to convert C language structures into their equivalent structures in Go. This will greatly help me with the code involving interactions with maps.

**Reply by lmb** [reply, marked as answer]
Have you tried passing `-type A -type B` to `bpf2go`? That should make it generate both types.

**Reply by binw666** [reply]
> Have you tried passing `-type A -type B` to `bpf2go`? That should make it generate both types.

Thank you very much for your response. I have not tried passing `-type A -type B` to `bpf2go`. I apologize for my oversight. After initiating the discussion, I saw the relevant instructions in bpf2go. After passing `-type A -type B` to `bpf2go`, the corresponding Go code is as follows :
```go
type bpfA struct {
	A1 uint32
	A2 uint8
	_  [3]byte
}

type bpfB struct {
	B1 uint16
	B2 uint8
	_  [1]byte
}

type bpfValue struct {
	Type  uint32
	Value struct{ A bpfA }
}
```
Using this method, I can obtain the struct I need. 

> Have you tried passing `-type A -type B` to `bpf2go`? That should make it generate both types.
